notes/compilation-and-linkage.md

```diff
@@ -0,0 +1,729 @@
+
+```C++
+/*
+#ifndef HEADER_FILE_NAME (All Caps is the convention)
+#define HEADER_FILE_NAME
+
+// Your code
+
+#endif
+*/
+#ifndef POINT_H
+#define POINT_H
+
+#define SHIFT_X = 4;
+#define SHIFT_Y = 5;
+
+#include <iostream>
+#include <string>
+
+struct point {
+    std::string name;
+    int x;
+    int y;
+};
+
+point DefaultTranslate();
+
+point Translate(std::string name, int xShift, int yShift);
+
+void printString(point p);
+
+#endif
+```
+<div style="text-align: center">
+
+Figure 7 - The Header Guard
+
+</div>
+
+This allows for header files, or any C/C++ file for that matter, to be copied only once per preprocessed file. Another way to do this is with the `#pragma` directive (This isn't common practice for LOST, so don't do it in HSL). It only requires a single declaration at the top of your program
+
+<div style="text-align: center">
+
+`#pragma once`
+
+</div>
+
+Anyway, with all that in mind, the preprocessor will operate exactly the same, but can now copy and paste `.hpp` files into `.cpp` files. You can imagine what `pointMain.cpp` looks like in a C++ preprocessor
+
+```C++
+// code for iostream
+// code for string
+
+struct point {
+    std::string name;
+    int x;
+    int y;
+};
+
+point DefaultTranslate();
+
+point Translate(std::string name, int xShift, int yShift);
+
+void printString(point p);
+
+int main() {
+    point myPoint = {"myPoint", 3, 10};
+    myPoint = Translate(myPoint.name, 0, 0);
+    printString();
+}
+/* Output:
+ * Coordinate Points for myPoint: (7, 15)
+**/
+```
+<div style="text-align: center">
+
+Figure 8 - C++ Preprocessing with `.hpp` files
+
+</div>
+
+But wait, how do we know what the function definitions are? Well, you'll have to find out in the next steps!
+
+## Compilation
+
+After you make your drawings of all the fundamental parts to your bridge, you need to tell your manufacturers how to create each part. You need to tell them in a way that will make sense to them, as engineering and manufacturing will use different terminology to describe the same thing. The compilation achieves this between the coder and the computer by translating code into instructions that are understood to the computer.
+
+After the preprocessor runs, the files we end up with are precursors to what is called a *translation unit*. This is a file that at minimum defines all entities used within a single file, and for the purposes of the coder, the *translation unit* that we understand is what is in the preprocessed file. It is everything that would be in each `.c/.cpp` file after the preprocessor runs.
+
+After the generation of each translation unit, the translation units are then processed a few more times, but the most notable intermediate processing that occurs is the translation of the files from C/C++ to Assembly code, or .asm. These are very basic instructions that talk directly to the registers of your computer how to run your code (You'll get to see it if you look it up or if you have/will take CSE 351).
+
```

Comment:
This section is solid - gives a good definition of each essential step without going into the weeds